<!DOCTYPE html>
<html><head><link href="css/default.css" rel="stylesheet" type="text/css"><script src="js/jquery.min.js" type="text/javascript"></script><script src="js/page_effects.js" type="text/javascript"></script><title>clj-biosequence.genbank documentation</title></head><body><div id="header"><h2>Generated by <a href="https://github.com/weavejester/codox">Codox</a></h2><h1><a href="index.html">Clj-biosequence 0.1.3-SNAPSHOT API documentation</a></h1></div><div class="sidebar" id="namespaces"><h3><span>Namespaces</span></h3><ul><li><a href="clj-biosequence.blast.html"><span>clj-biosequence.blast</span></a></li><li><a href="clj-biosequence.core.html"><span>clj-biosequence.core</span></a></li><li class="current"><a href="clj-biosequence.genbank.html"><span>clj-biosequence.genbank</span></a></li><li><a href="clj-biosequence.signalp.html"><span>clj-biosequence.signalp</span></a></li><li><a href="clj-biosequence.utilities.html"><span>clj-biosequence.utilities</span></a></li></ul></div><div class="sidebar" id="vars"><h3>Public Vars</h3><ul><li><a href="clj-biosequence.genbank.html#var--%3EgenbankFeature"><span>-&gt;genbankFeature</span></a></li><li><a href="clj-biosequence.genbank.html#var--%3EgenbankFile"><span>-&gt;genbankFile</span></a></li><li><a href="clj-biosequence.genbank.html#var--%3EgenbankInterval"><span>-&gt;genbankInterval</span></a></li><li><a href="clj-biosequence.genbank.html#var--%3EgenbankSequence"><span>-&gt;genbankSequence</span></a></li><li><a href="clj-biosequence.genbank.html#var--%3EgenbankStore"><span>-&gt;genbankStore</span></a></li><li><a href="clj-biosequence.genbank.html#var-comp%3F"><span>comp?</span></a></li><li><a href="clj-biosequence.genbank.html#var-end"><span>end</span></a></li><li><a href="clj-biosequence.genbank.html#var-feature-seq"><span>feature-seq</span></a></li><li><a href="clj-biosequence.genbank.html#var-feature-type"><span>feature-type</span></a></li><li><a href="clj-biosequence.genbank.html#var-gb-locus"><span>gb-locus</span></a></li><li><a href="clj-biosequence.genbank.html#var-get-feature-sequence"><span>get-feature-sequence</span></a></li><li><a href="clj-biosequence.genbank.html#var-get-interval-sequence"><span>get-interval-sequence</span></a></li><li><a href="clj-biosequence.genbank.html#var-index-genbank-file"><span>index-genbank-file</span></a></li><li><a href="clj-biosequence.genbank.html#var-interval-seq"><span>interval-seq</span></a></li><li><a href="clj-biosequence.genbank.html#var-map-%3EgenbankFeature"><span>map-&gt;genbankFeature</span></a></li><li><a href="clj-biosequence.genbank.html#var-map-%3EgenbankFile"><span>map-&gt;genbankFile</span></a></li><li><a href="clj-biosequence.genbank.html#var-map-%3EgenbankInterval"><span>map-&gt;genbankInterval</span></a></li><li><a href="clj-biosequence.genbank.html#var-map-%3EgenbankSequence"><span>map-&gt;genbankSequence</span></a></li><li><a href="clj-biosequence.genbank.html#var-map-%3EgenbankStore"><span>map-&gt;genbankStore</span></a></li><li><a href="clj-biosequence.genbank.html#var-qualifier-extract"><span>qualifier-extract</span></a></li><li><a href="clj-biosequence.genbank.html#var-start"><span>start</span></a></li><li><a href="clj-biosequence.genbank.html#var-wget-genbank-search"><span>wget-genbank-search</span></a></li><li><a href="clj-biosequence.genbank.html#var-with-wget-genbank-sequence"><span>with-wget-genbank-sequence</span></a></li></ul></div><div class="namespace-docs" id="content"><h2>clj-biosequence.genbank documentation</h2><pre class="doc"></pre><div class="public" id="var--%3EgenbankFeature"><h3>-&gt;genbankFeature</h3><div class="usage"><code>(-&gt;genbankFeature src)</code></div><pre class="doc">Positional factory function for class clj_biosequence.genbank.genbankFeature.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L71">Source</a></div></div><div class="public" id="var--%3EgenbankFile"><h3>-&gt;genbankFile</h3><div class="usage"><code>(-&gt;genbankFile file)</code></div><pre class="doc">Positional factory function for class clj_biosequence.genbank.genbankFile.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L250">Source</a></div></div><div class="public" id="var--%3EgenbankInterval"><h3>-&gt;genbankInterval</h3><div class="usage"><code>(-&gt;genbankInterval src)</code></div><pre class="doc">Positional factory function for class clj_biosequence.genbank.genbankInterval.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L17">Source</a></div></div><div class="public" id="var--%3EgenbankSequence"><h3>-&gt;genbankSequence</h3><div class="usage"><code>(-&gt;genbankSequence src)</code></div><pre class="doc">Positional factory function for class clj_biosequence.genbank.genbankSequence.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L146">Source</a></div></div><div class="public" id="var--%3EgenbankStore"><h3>-&gt;genbankStore</h3><div class="usage"><code>(-&gt;genbankStore file)</code></div><pre class="doc">Positional factory function for class clj_biosequence.genbank.genbankStore.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L268">Source</a></div></div><div class="public" id="var-comp%3F"><h3>comp?</h3><div class="usage"><code>(comp? gb-interval)</code></div><pre class="doc">Is a genbankInterval complementary?
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L45">Source</a></div></div><div class="public" id="var-end"><h3>end</h3><div class="usage"><code>(end gb-interval)</code></div><pre class="doc">End index of a genbankInterval.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L32">Source</a></div></div><div class="public" id="var-feature-seq"><h3>feature-seq</h3><div class="usage"><code>(feature-seq gbseq)</code></div><pre class="doc">Returns a lazy list of features from a genbankSequence.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L228">Source</a></div></div><div class="public" id="var-feature-type"><h3>feature-type</h3><div class="usage"><code>(feature-type feature)</code></div><pre class="doc">Returns the feature key. For example: protein, Region, Site, CDS etc.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L73">Source</a></div></div><div class="public" id="var-gb-locus"><h3>gb-locus</h3><div class="usage"><code>(gb-locus gbseq)</code></div><pre class="doc">Returns the locus of a genbankSequence.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L223">Source</a></div></div><div class="public" id="var-get-feature-sequence"><h3>get-feature-sequence</h3><div class="usage"><code>(get-feature-sequence gb-feat gbseq)</code></div><pre class="doc">Returns a fastaSequence object containing the sequence specified in a genbankFeature
object from a genbankSequence object. Designed for applying intervals to the sequence
entry they originate from.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L123">Source</a></div></div><div class="public" id="var-get-interval-sequence"><h3>get-interval-sequence</h3><div class="usage"><code>(get-interval-sequence gb-interval gb-sequence)</code></div><pre class="doc">Returns a fastaSequence object containing the sequence specified in the
genbankInterval from a genbankSequence. Designed for getting fastasequences
by applying  genbankInterval to the sequence entry it originates from.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L53">Source</a></div></div><div class="public" id="var-index-genbank-file"><h3>index-genbank-file</h3><div class="usage"><code>(index-genbank-file genbankfile)</code><code>(index-genbank-file genbankfile memory)</code></div><pre class="doc">Indexes a genbankFile object and returns a uniprotStore object. Duplicate 
primary accessions in the file will cause a 'Unique index or primary key 
violation'.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L270">Source</a></div></div><div class="public" id="var-interval-seq"><h3>interval-seq</h3><div class="usage"><code>(interval-seq gb-feature)</code></div><pre class="doc">Returns a non-lazy list of intervals in a feature. It also calculates a frame
for each interval so that individual translations of a DNA interval provide the
correct protein sequence. This value is accessed through the :frame keyword for
each interval. These values will have no meaning if the sequence is a protein.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L94">Source</a></div></div><div class="public" id="var-map-%3EgenbankFeature"><h3>map-&gt;genbankFeature</h3><div class="usage"><code>(map-&gt;genbankFeature m__5665__auto__)</code></div><pre class="doc">Factory function for class clj_biosequence.genbank.genbankFeature, taking a map of keywords to field values.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L71">Source</a></div></div><div class="public" id="var-map-%3EgenbankFile"><h3>map-&gt;genbankFile</h3><div class="usage"><code>(map-&gt;genbankFile m__5665__auto__)</code></div><pre class="doc">Factory function for class clj_biosequence.genbank.genbankFile, taking a map of keywords to field values.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L250">Source</a></div></div><div class="public" id="var-map-%3EgenbankInterval"><h3>map-&gt;genbankInterval</h3><div class="usage"><code>(map-&gt;genbankInterval m__5665__auto__)</code></div><pre class="doc">Factory function for class clj_biosequence.genbank.genbankInterval, taking a map of keywords to field values.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L17">Source</a></div></div><div class="public" id="var-map-%3EgenbankSequence"><h3>map-&gt;genbankSequence</h3><div class="usage"><code>(map-&gt;genbankSequence m__5665__auto__)</code></div><pre class="doc">Factory function for class clj_biosequence.genbank.genbankSequence, taking a map of keywords to field values.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L146">Source</a></div></div><div class="public" id="var-map-%3EgenbankStore"><h3>map-&gt;genbankStore</h3><div class="usage"><code>(map-&gt;genbankStore m__5665__auto__)</code></div><pre class="doc">Factory function for class clj_biosequence.genbank.genbankStore, taking a map of keywords to field values.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L268">Source</a></div></div><div class="public" id="var-qualifier-extract"><h3>qualifier-extract</h3><div class="usage"><code>(qualifier-extract feature element)</code></div><pre class="doc">Takes a genbankFeature and returns the value specified by 'element'. For 
instance to retrieve the organism name from a 'source' feature arguments 
would be the feature object and 'organism'.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L80">Source</a></div></div><div class="public" id="var-start"><h3>start</h3><div class="usage"><code>(start gb-interval)</code></div><pre class="doc">Start index of a genbankInterval.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L19">Source</a></div></div><div class="public" id="var-wget-genbank-search"><h3>wget-genbank-search</h3><div class="usage"><code>(wget-genbank-search term db)</code><code>(wget-genbank-search term db restart key)</code></div><pre class="doc">Returns a lazy list of result ids from NCBI for a particular search term and
database. Search term syntax is the same as is used at the NCBI. For example,
to retreive all Schistosoma mansoni protein sequences the 'db' argument would
be 'protein' and the search term 'txid6183[Organism:noexp]'. Database arguments
are restricted to the following keyword arguments and any other value will cause
an error:
:protein    - a collection of sequences from several sources, including 
              translations from annotated coding regions in GenBank, RefSeq 
              and TPA, as well as records from SwissProt, PIR, PRF, and PDB.
:nucest     - a collection of short single-read transcript sequences from GenBank.
:nuccore    - a collection of sequences from several sources, including GenBank, 
              RefSeq, TPA and PDB (same as :nucleotide).
:nucgss     - a collection of unannotated short single-read primarily genomic
              sequences from GenBank including random survey sequences clone-end
              sequences and exon-trapped sequences.
:popset     - a set of DNA sequences that have been collected to analyse the 
              evolutionary relatedness of a population. The population could 
              originate from different members of the same species, or from 
              organisms from different species. 

Note that when retrieving popset entries multiple sequences are returned from a
single accession number. Returns an empty list if no matches found.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L344">Source</a></div></div><div class="public" id="var-with-wget-genbank-sequence"><h3>with-wget-genbank-sequence</h3><h4 class="macro">macro</h4><div class="usage"><code>(with-wget-genbank-sequence [handle accessions db rettype] &amp; code)</code></div><pre class="doc">Returns a 'semi-lazy' (1000 sequences are fetched at a time) list of 
genbankSequences corresponding to the 'accessions' list argument. Accepts
any identification that genbank accepts but will return an exception 
(status code 400) if the id is not recognised. Can return genbankSequences 
or fastaSequence objects depending whether the 'class' argument is :xml or
:fasta respectively. The 'db' argument has the same format and allowed values
as the wget-genbank-search function.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/genbank.clj#L309">Source</a></div></div></div></body></html>