<!DOCTYPE html>
<html><head><link href="css/default.css" rel="stylesheet" type="text/css"><script src="js/jquery.min.js" type="text/javascript"></script><script src="js/page_effects.js" type="text/javascript"></script><title>clj-biosequence.core documentation</title></head><body><div id="header"><h2>Generated by <a href="https://github.com/weavejester/codox">Codox</a></h2><h1><a href="index.html">Clj-biosequence 0.1.3-SNAPSHOT API documentation</a></h1></div><div class="sidebar" id="namespaces"><h3><span>Namespaces</span></h3><ul><li><a href="clj-biosequence.blast.html"><span>clj-biosequence.blast</span></a></li><li class="current"><a href="clj-biosequence.core.html"><span>clj-biosequence.core</span></a></li><li><a href="clj-biosequence.genbank.html"><span>clj-biosequence.genbank</span></a></li><li><a href="clj-biosequence.signalp.html"><span>clj-biosequence.signalp</span></a></li><li><a href="clj-biosequence.utilities.html"><span>clj-biosequence.utilities</span></a></li></ul></div><div class="sidebar" id="vars"><h3>Public Vars</h3><ul><li><a href="clj-biosequence.core.html#var--%3EfastaFile"><span>-&gt;fastaFile</span></a></li><li><a href="clj-biosequence.core.html#var--%3EfastaSequence"><span>-&gt;fastaSequence</span></a></li><li><a href="clj-biosequence.core.html#var--%3EfastaStore"><span>-&gt;fastaStore</span></a></li><li><a href="clj-biosequence.core.html#var-accession"><span>accession</span></a></li><li><a href="clj-biosequence.core.html#var-accessions"><span>accessions</span></a></li><li><a href="clj-biosequence.core.html#var-biosequence-seq-file"><span>biosequence-seq-file</span></a></li><li><a href="clj-biosequence.core.html#var-codon-tables"><span>codon-tables</span></a></li><li><a href="clj-biosequence.core.html#var-created"><span>created</span></a></li><li><a href="clj-biosequence.core.html#var-database"><span>database</span></a></li><li><a href="clj-biosequence.core.html#var-def-line"><span>def-line</span></a></li><li><a href="clj-biosequence.core.html#var-fasta-seq-string"><span>fasta-seq-string</span></a></li><li><a href="clj-biosequence.core.html#var-fasta-string"><span>fasta-string</span></a></li><li><a href="clj-biosequence.core.html#var-file-path"><span>file-path</span></a></li><li><a href="clj-biosequence.core.html#var-id-convert"><span>id-convert</span></a></li><li><a href="clj-biosequence.core.html#var-index-fasta-file"><span>index-fasta-file</span></a></li><li><a href="clj-biosequence.core.html#var-index-file-name"><span>index-file-name</span></a></li><li><a href="clj-biosequence.core.html#var-init-fasta-file"><span>init-fasta-file</span></a></li><li><a href="clj-biosequence.core.html#var-init-fasta-sequence"><span>init-fasta-sequence</span></a></li><li><a href="clj-biosequence.core.html#var-init-in-mem-store"><span>init-in-mem-store</span></a></li><li><a href="clj-biosequence.core.html#var-init-store"><span>init-store</span></a></li><li><a href="clj-biosequence.core.html#var-load-fasta-store"><span>load-fasta-store</span></a></li><li><a href="clj-biosequence.core.html#var-make-db-connection"><span>make-db-connection</span></a></li><li><a href="clj-biosequence.core.html#var-make-mem-db-connection"><span>make-mem-db-connection</span></a></li><li><a href="clj-biosequence.core.html#var-map-%3EfastaFile"><span>map-&gt;fastaFile</span></a></li><li><a href="clj-biosequence.core.html#var-map-%3EfastaSequence"><span>map-&gt;fastaSequence</span></a></li><li><a href="clj-biosequence.core.html#var-map-%3EfastaStore"><span>map-&gt;fastaStore</span></a></li><li><a href="clj-biosequence.core.html#var-modified"><span>modified</span></a></li><li><a href="clj-biosequence.core.html#var-org-scientific-name"><span>org-scientific-name</span></a></li><li><a href="clj-biosequence.core.html#var-protein%3F"><span>protein?</span></a></li><li><a href="clj-biosequence.core.html#var-revcom-dna-string"><span>revcom-dna-string</span></a></li><li><a href="clj-biosequence.core.html#var-sequence-string"><span>sequence-string</span></a></li><li><a href="clj-biosequence.core.html#var-six-frame-translation"><span>six-frame-translation</span></a></li><li><a href="clj-biosequence.core.html#var-taxid"><span>taxid</span></a></li><li><a href="clj-biosequence.core.html#var-taxonomy"><span>taxonomy</span></a></li><li><a href="clj-biosequence.core.html#var-translate-biosequence"><span>translate-biosequence</span></a></li><li><a href="clj-biosequence.core.html#var-update-object-by-accession"><span>update-object-by-accession</span></a></li><li><a href="clj-biosequence.core.html#var-version"><span>version</span></a></li><li><a href="clj-biosequence.core.html#var-with-biosequences"><span>with-biosequences</span></a></li><li><a href="clj-biosequence.core.html#var-with-connection-to-store"><span>with-connection-to-store</span></a></li><li><a href="clj-biosequence.core.html#var-with-fasta-file"><span>with-fasta-file</span></a></li></ul></div><div class="namespace-docs" id="content"><h2>clj-biosequence.core documentation</h2><pre class="doc"></pre><div class="public" id="var--%3EfastaFile"><h3>-&gt;fastaFile</h3><div class="usage"><code>(-&gt;fastaFile file type)</code></div><pre class="doc">Positional factory function for class clj_biosequence.core.fastaFile.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L317">Source</a></div></div><div class="public" id="var--%3EfastaSequence"><h3>-&gt;fastaSequence</h3><div class="usage"><code>(-&gt;fastaSequence accession description type sequence)</code></div><pre class="doc">Positional factory function for class clj_biosequence.core.fastaSequence.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L242">Source</a></div></div><div class="public" id="var--%3EfastaStore"><h3>-&gt;fastaStore</h3><div class="usage"><code>(-&gt;fastaStore file type)</code></div><pre class="doc">Positional factory function for class clj_biosequence.core.fastaStore.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L351">Source</a></div></div><div class="public" id="var-accession"><h3>accession</h3><div class="usage"><code>(accession this)</code></div><pre class="doc">Returns the accession of a biosequence.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-accessions"><h3>accessions</h3><div class="usage"><code>(accessions this)</code></div><pre class="doc">Returns a list of strings describing the accessions of a biosequence object.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-biosequence-seq-file"><h3>biosequence-seq-file</h3><div class="usage"><code>(biosequence-seq-file this rdr)</code></div><pre class="doc">Returns a lazy sequence of biosequences from a biosequence file object.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-codon-tables"><h3>codon-tables</h3><div class="usage"><code>(codon-tables key)</code></div><pre class="doc">Returns a codon table suitable for use as an argument in the functions 
'translate-biosequence' and 'six-frame-translation'. Takes a keyword argument
 to denote different codon tables. So far only provides the standard (:standard)
 codon table but further are planned.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L491">Source</a></div></div><div class="public" id="var-created"><h3>created</h3><div class="usage"><code>(created this)</code></div><pre class="doc">Date sequence created.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-database"><h3>database</h3><div class="usage"><code>(database this)</code></div><pre class="doc">Returns the database that contains the sequence.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-def-line"><h3>def-line</h3><div class="usage"><code>(def-line this)</code></div><pre class="doc">Returns a description for a biosequence object.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-fasta-seq-string"><h3>fasta-seq-string</h3><div class="usage"><code>(fasta-seq-string string type)</code></div><pre class="doc">Returns a non-lazy list of fastaSequence objects from a string.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L342">Source</a></div></div><div class="public" id="var-fasta-string"><h3>fasta-string</h3><div class="usage"><code>(fasta-string this)</code></div><pre class="doc">Returns the biosequence as a string in fasta format.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-file-path"><h3>file-path</h3><div class="usage"><code>(file-path this)</code></div><pre class="doc">Returns the path of a biosequence file.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-id-convert"><h3>id-convert</h3><div class="usage"><code>(id-convert ids from to email)</code></div><pre class="doc">Takes either a single accession or a list of accessions and returns a hash-map
mapping the accession numbers to the corresponding identification number in the 
specified 'to' database. 'From' database also needs to be specified. If not 
found returns an empty hash-map. Uses the Uniprot id mapping utility and a list 
of supported databases is supplied at http://www.uniprot.org/faq/28#id_mapping_examples.
Some common mappings include:
DB Name                  Abbreviation     Direction
UniProtKB AC/ID	    ACC+ID	     from
UniProtKB AC	            ACC              to
EMBL/GenBank/DDBJ	    EMBL_ID	     both
EMBL/GenBank/DDBJ CDS    EMBL	     both
Entrez Gene (GeneID)     P_ENTREZGENEID   both
GI number	            P_GI	     both
RefSeq Protein	    P_REFSEQ_AC	     both
RefSeq Nucleotide	    REFSEQ_NT_ID     both
WormBase	            WORMBASE_ID	     both

There is a 100,000 limit on accessions in a single query imposed by Uniprot.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L377">Source</a></div></div><div class="public" id="var-index-fasta-file"><h3>index-fasta-file</h3><div class="usage"><code>(index-fasta-file fastafile)</code><code>(index-fasta-file fastafile memory)</code></div><pre class="doc">Indexes a fastaFile object and returns a fastaStore object.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L353">Source</a></div></div><div class="public" id="var-index-file-name"><h3>index-file-name</h3><div class="usage"><code>(index-file-name file)</code></div><pre class="doc">Returns a path suitable for a persistent store.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L226">Source</a></div></div><div class="public" id="var-init-fasta-file"><h3>init-fasta-file</h3><div class="usage"><code>(init-fasta-file path type)</code></div><pre class="doc">Initialises fasta protein file. Accession numbers and description are processed by splitting 
the string on the first space, the accession being the first value and description the second.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L332">Source</a></div></div><div class="public" id="var-init-fasta-sequence"><h3>init-fasta-sequence</h3><div class="usage"><code>(init-fasta-sequence accession description type sequence)</code></div><pre class="doc">Returns a fastaSequence object with the specified information.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L244">Source</a></div></div><div class="public" id="var-init-in-mem-store"><h3>init-in-mem-store</h3><div class="usage"><code>(init-in-mem-store store)</code></div><pre class="doc">Initialises an in-memory permanent store
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L188">Source</a></div></div><div class="public" id="var-init-store"><h3>init-store</h3><div class="usage"><code>(init-store store)</code></div><pre class="doc">Initialises a permanent store.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L170">Source</a></div></div><div class="public" id="var-load-fasta-store"><h3>load-fasta-store</h3><div class="usage"><code>(load-fasta-store dir type)</code></div><pre class="doc">Loads a fastaStore.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L363">Source</a></div></div><div class="public" id="var-make-db-connection"><h3>make-db-connection</h3><div class="usage"><code>(make-db-connection file exists)</code></div><pre class="doc">Takes a file path and returns a specification for a database based on that file.
Exists is a boolean and specifies whether an error is signalled if file already exists.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L148">Source</a></div></div><div class="public" id="var-make-mem-db-connection"><h3>make-mem-db-connection</h3><div class="usage"><code>(make-mem-db-connection)</code></div><pre class="doc">Returns a in-memory database specification.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L160">Source</a></div></div><div class="public" id="var-map-%3EfastaFile"><h3>map-&gt;fastaFile</h3><div class="usage"><code>(map-&gt;fastaFile m__5665__auto__)</code></div><pre class="doc">Factory function for class clj_biosequence.core.fastaFile, taking a map of keywords to field values.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L317">Source</a></div></div><div class="public" id="var-map-%3EfastaSequence"><h3>map-&gt;fastaSequence</h3><div class="usage"><code>(map-&gt;fastaSequence m__5665__auto__)</code></div><pre class="doc">Factory function for class clj_biosequence.core.fastaSequence, taking a map of keywords to field values.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L242">Source</a></div></div><div class="public" id="var-map-%3EfastaStore"><h3>map-&gt;fastaStore</h3><div class="usage"><code>(map-&gt;fastaStore m__5665__auto__)</code></div><pre class="doc">Factory function for class clj_biosequence.core.fastaStore, taking a map of keywords to field values.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L351">Source</a></div></div><div class="public" id="var-modified"><h3>modified</h3><div class="usage"><code>(modified this)</code></div><pre class="doc">Date sequence modified.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-org-scientific-name"><h3>org-scientific-name</h3><div class="usage"><code>(org-scientific-name this)</code></div><pre class="doc">Returns the scientific name of the organism possessing the sequence as a string.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-protein%3F"><h3>protein?</h3><div class="usage"><code>(protein? this)</code></div><pre class="doc">Returns true if a protein and false otherwise.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-revcom-dna-string"><h3>revcom-dna-string</h3><div class="usage"><code>(revcom-dna-string string)</code></div><pre class="doc">Provides the reverse, complement of a string representing a DNA sequence. 
Converts all 'U's to 'T's and upcases the string.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L537">Source</a></div></div><div class="public" id="var-sequence-string"><h3>sequence-string</h3><div class="usage"><code>(sequence-string this)</code></div><pre class="doc">Returns the sequence of a biosequence as a string.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-six-frame-translation"><h3>six-frame-translation</h3><div class="usage"><code>(six-frame-translation nucleotide)</code><code>(six-frame-translation nucleotide table)</code></div><pre class="doc">Returns a lazy list of fastaSequence objects representing translations of
a nucleotide biosequence object in six frames.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L122">Source</a></div></div><div class="public" id="var-taxid"><h3>taxid</h3><div class="usage"><code>(taxid this)</code></div><pre class="doc">Returns the taxid of a sequence (Integer).
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-taxonomy"><h3>taxonomy</h3><div class="usage"><code>(taxonomy this)</code></div><pre class="doc">Returns the lineage from a uniprot as a list of strings. Strings in order from kingdom to species.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-translate-biosequence"><h3>translate-biosequence</h3><div class="usage"><code>(translate-biosequence nucleotide frame)</code><code>(translate-biosequence nucleotide frame table)</code></div><pre class="doc">Returns a fastaSequence object corresponding to the protein translation 
of the sequence in the specified frame.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L114">Source</a></div></div><div class="public" id="var-update-object-by-accession"><h3>update-object-by-accession</h3><div class="usage"><code>(update-object-by-accession accession &amp; args)</code></div><pre class="doc">Takes an accession number and key value pairs. If the object exists in
the current store the object with the corresponding accession number
will be updated with the key value pairs and saved. Throws an exception
if a corresponding object is not found in the store.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L216">Source</a></div></div><div class="public" id="var-version"><h3>version</h3><div class="usage"><code>(version this)</code></div><pre class="doc">Version of the sequence (Integer).
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/#L">Source</a></div></div><div class="public" id="var-with-biosequences"><h3>with-biosequences</h3><h4 class="macro">macro</h4><div class="usage"><code>(with-biosequences [handle sink] &amp; body)</code></div><pre class="doc">Provides a handle to a lazy list of biosequences in a biosequence store or
file. Also works with collections of biosequences.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L20">Source</a></div></div><div class="public" id="var-with-connection-to-store"><h3>with-connection-to-store</h3><h4 class="macro">macro</h4><div class="usage"><code>(with-connection-to-store [store] &amp; code)</code></div><pre class="doc">Provides a connection to a biosequence store.
</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L13">Source</a></div></div><div class="public" id="var-with-fasta-file"><h3>with-fasta-file</h3><h4 class="macro">macro</h4><div class="usage"><code>(with-fasta-file [handle sink] &amp; code)</code></div><pre class="doc">Provides a handle to a temporary fasta file containing the biosequences
contained in a biosequence file, store or collection.</pre><div class="src-link"><a href="https://github.com/s312569/clj-biosequence/blob/master/src/clj_biosequence/core.clj#L41">Source</a></div></div></div></body></html>