<!DOCTYPE html>
<html><head><link href="css/default.css" rel="stylesheet" type="text/css"><script src="js/jquery.min.js" type="text/javascript"></script><script src="js/page_effects.js" type="text/javascript"></script><title>clj-biosequence.uniprot documentation</title></head><body><div id="header"><h2>Generated by <a href="https://github.com/weavejester/codox">Codox</a></h2><h1><a href="index.html">Clj-biosequence 0.1.2-SNAPSHOT API documentation</a></h1></div><div class="sidebar" id="namespaces"><h3><span>Namespaces</span></h3><ul><li><a href="clj-biosequence.blast.html"><span>clj-biosequence.blast</span></a></li><li><a href="clj-biosequence.core.html"><span>clj-biosequence.core</span></a></li><li><a href="clj-biosequence.genbank.html"><span>clj-biosequence.genbank</span></a></li><li><a href="clj-biosequence.signalp.html"><span>clj-biosequence.signalp</span></a></li><li class="current"><a href="clj-biosequence.uniprot.html"><span>clj-biosequence.uniprot</span></a></li><li><a href="clj-biosequence.utilities.html"><span>clj-biosequence.utilities</span></a></li></ul></div><div class="sidebar" id="vars"><h3>Public Vars</h3><ul><li><a href="clj-biosequence.uniprot.html#var--%3EuniprotProtein"><span>-&gt;uniprotProtein</span></a></li><li><a href="clj-biosequence.uniprot.html#var--%3EuniprotStore"><span>-&gt;uniprotStore</span></a></li><li><a href="clj-biosequence.uniprot.html#var--%3EuniprotXmlFile"><span>-&gt;uniprotXmlFile</span></a></li><li><a href="clj-biosequence.uniprot.html#var-alternative-products"><span>alternative-products</span></a></li><li><a href="clj-biosequence.uniprot.html#var-amino-acids"><span>amino-acids</span></a></li><li><a href="clj-biosequence.uniprot.html#var-citation"><span>citation</span></a></li><li><a href="clj-biosequence.uniprot.html#var-comments"><span>comments</span></a></li><li><a href="clj-biosequence.uniprot.html#var-db-references"><span>db-references</span></a></li><li><a href="clj-biosequence.uniprot.html#var-existence"><span>existence</span></a></li><li><a href="clj-biosequence.uniprot.html#var-features"><span>features</span></a></li><li><a href="clj-biosequence.uniprot.html#var-gene"><span>gene</span></a></li><li><a href="clj-biosequence.uniprot.html#var-gene-location"><span>gene-location</span></a></li><li><a href="clj-biosequence.uniprot.html#var-index-uniprotxml-file"><span>index-uniprotxml-file</span></a></li><li><a href="clj-biosequence.uniprot.html#var-init-uniprotxml-file"><span>init-uniprotxml-file</span></a></li><li><a href="clj-biosequence.uniprot.html#var-interactions"><span>interactions</span></a></li><li><a href="clj-biosequence.uniprot.html#var-keywords"><span>keywords</span></a></li><li><a href="clj-biosequence.uniprot.html#var-load-uniprot-store"><span>load-uniprot-store</span></a></li><li><a href="clj-biosequence.uniprot.html#var-map-%3EuniprotProtein"><span>map-&gt;uniprotProtein</span></a></li><li><a href="clj-biosequence.uniprot.html#var-map-%3EuniprotStore"><span>map-&gt;uniprotStore</span></a></li><li><a href="clj-biosequence.uniprot.html#var-map-%3EuniprotXmlFile"><span>map-&gt;uniprotXmlFile</span></a></li><li><a href="clj-biosequence.uniprot.html#var-mass-spectroscopy"><span>mass-spectroscopy</span></a></li><li><a href="clj-biosequence.uniprot.html#var-nomenclature"><span>nomenclature</span></a></li><li><a href="clj-biosequence.uniprot.html#var-organism-name"><span>organism-name</span></a></li><li><a href="clj-biosequence.uniprot.html#var-prot-name"><span>prot-name</span></a></li><li><a href="clj-biosequence.uniprot.html#var-subcellular-location"><span>subcellular-location</span></a></li><li><a href="clj-biosequence.uniprot.html#var-wget-uniprot-search"><span>wget-uniprot-search</span></a></li><li><a href="clj-biosequence.uniprot.html#var-wget-uniprot-sequence"><span>wget-uniprot-sequence</span></a></li></ul></div><div class="namespace-docs" id="content"><h2>clj-biosequence.uniprot documentation</h2><pre class="doc"></pre><div class="public" id="var--%3EuniprotProtein"><h3>-&gt;uniprotProtein</h3><div class="usage"><code>(-&gt;uniprotProtein accession src)</code></div><pre class="doc">Positional factory function for class clj_biosequence.uniprot.uniprotProtein.
</pre></div><div class="public" id="var--%3EuniprotStore"><h3>-&gt;uniprotStore</h3><div class="usage"><code>(-&gt;uniprotStore file)</code></div><pre class="doc">Positional factory function for class clj_biosequence.uniprot.uniprotStore.
</pre></div><div class="public" id="var--%3EuniprotXmlFile"><h3>-&gt;uniprotXmlFile</h3><div class="usage"><code>(-&gt;uniprotXmlFile file)</code></div><pre class="doc">Positional factory function for class clj_biosequence.uniprot.uniprotXmlFile.
</pre></div><div class="public" id="var-alternative-products"><h3>alternative-products</h3><div class="usage"><code>(alternative-products uniprot)</code></div><pre class="doc">Returns a list of maps describing alternative products of a uniprot. Each map has the keys - :event, :isoform-id, :isoform-name and :comments. All of which are lists of strings except :comments which is a string.
</pre></div><div class="public" id="var-amino-acids"><h3>amino-acids</h3><div class="usage"><code>(amino-acids uniprot)</code></div><pre class="doc">Returns a map with the keys - :mass, :checksum, :modified, :version, :amino-acids . All are strings except :mass, which is a long, and :version, which is an integer.
</pre></div><div class="public" id="var-citation"><h3>citation</h3><div class="usage"><code>(citation uniprot)</code></div><pre class="doc">Returns a list of citations from a uniprot. Each citation is a map with the keys - :country, :last (page), :date, :pubmed, :institute, :name (journal), :first (page), :title, :city, :scope, :type, :consortium, :number, :authors, :source, :editors, :publisher, :volume, :db. All are strings, except for :scope, :editor, :strain and :author, which are lists of strings. The value of :source is a map with the keys - :tissue, :transposons, :plasmid and :strain which are all lists of strings.
</pre></div><div class="public" id="var-comments"><h3>comments</h3><div class="usage"><code>(comments uniprot)</code></div><pre class="doc">Returns a map of comments describing a uniprot. Keys - :similarity and :function both of which have strings as values.
</pre></div><div class="public" id="var-db-references"><h3>db-references</h3><div class="usage"><code>(db-references uniprot)</code></div><pre class="doc">Returns a list of maps describing database cross-references of a uniprot. Each map has the keys :data, which has a map as a value with a range of different keys corresponding to the xml file, :db, a string describing the database, :id, a string of the id of the cross-ref.
</pre></div><div class="public" id="var-existence"><h3>existence</h3><div class="usage"><code>(existence uniprot)</code></div><pre class="doc">Returns a list of strings describing the evidence for the existence of a uniprot.
</pre></div><div class="public" id="var-features"><h3>features</h3><div class="usage"><code>(features uniprot)</code></div><pre class="doc">A list of maps describing the features of a uniprot. Each map has the keys - :id, :description, :type, :begin, :end. All have string values except :begin and :end which are integers.
</pre></div><div class="public" id="var-gene"><h3>gene</h3><div class="usage"><code>(gene uniprot)</code></div><pre class="doc">Returns a map with gene information from a uniprot. Keys - :type and :name. Values are strings.
</pre></div><div class="public" id="var-gene-location"><h3>gene-location</h3><div class="usage"><code>(gene-location uniprot)</code></div><pre class="doc">Returns a map of gene location information from a uniprot.
</pre></div><div class="public" id="var-index-uniprotxml-file"><h3>index-uniprotxml-file</h3><div class="usage"><code>(index-uniprotxml-file uniprotfile)</code><code>(index-uniprotxml-file uniprotfile memory)</code></div><pre class="doc">Indexes a uniprotXmlFile and returns a uniprotStore object.
</pre></div><div class="public" id="var-init-uniprotxml-file"><h3>init-uniprotxml-file</h3><div class="usage"><code>(init-uniprotxml-file path)</code></div><pre class="doc">Initialises a uniprotXmlFile object.
</pre></div><div class="public" id="var-interactions"><h3>interactions</h3><div class="usage"><code>(interactions uniprot)</code></div><pre class="doc">Returns a list of maps describing interactions of a uniprot. Each map has the keys - :id, :label, :interactors, :experiments, :comments. All of whcih are lists of strings, except :experiments and :comments which are strings.
</pre></div><div class="public" id="var-keywords"><h3>keywords</h3><div class="usage"><code>(keywords uniprot)</code></div><pre class="doc">A map of uniprot keywords. Keys are the keyword id with a string value of the keyword.
</pre></div><div class="public" id="var-load-uniprot-store"><h3>load-uniprot-store</h3><div class="usage"><code>(load-uniprot-store dir)</code></div><pre class="doc">Loads a uniprotStore.
</pre></div><div class="public" id="var-map-%3EuniprotProtein"><h3>map-&gt;uniprotProtein</h3><div class="usage"><code>(map-&gt;uniprotProtein m__5665__auto__)</code></div><pre class="doc">Factory function for class clj_biosequence.uniprot.uniprotProtein, taking a map of keywords to field values.
</pre></div><div class="public" id="var-map-%3EuniprotStore"><h3>map-&gt;uniprotStore</h3><div class="usage"><code>(map-&gt;uniprotStore m__5665__auto__)</code></div><pre class="doc">Factory function for class clj_biosequence.uniprot.uniprotStore, taking a map of keywords to field values.
</pre></div><div class="public" id="var-map-%3EuniprotXmlFile"><h3>map-&gt;uniprotXmlFile</h3><div class="usage"><code>(map-&gt;uniprotXmlFile m__5665__auto__)</code></div><pre class="doc">Factory function for class clj_biosequence.uniprot.uniprotXmlFile, taking a map of keywords to field values.
</pre></div><div class="public" id="var-mass-spectroscopy"><h3>mass-spectroscopy</h3><div class="usage"><code>(mass-spectroscopy uniprot)</code></div><pre class="doc">Returns a list of maps describing the mass spectroscopy of a uniprot. Each map has the keys - :mass, :error, :method, :begin, :end and :comments. 
</pre></div><div class="public" id="var-nomenclature"><h3>nomenclature</h3><div class="usage"><code>(nomenclature uniprot)</code></div><pre class="doc">Returns a map of descriptions. Keys - :recommended, :alternate, :submitted, :allergen. Values of each is a list of maps with keys - :fullname, :shortname and :ecnumber. All values are strings except for :recommended which is map with the keys :fullname, :shortname and :ecnumber.
</pre></div><div class="public" id="var-organism-name"><h3>organism-name</h3><div class="usage"><code>(organism-name uniprot)</code></div><pre class="doc">Returns a map with keys - :scientific, :common, :full, :synonym, :abbreviation and :ncbi-taxid. Values are lists of names attributed to the organism, except :ncbi-taxid which is an integer.
</pre></div><div class="public" id="var-prot-name"><h3>prot-name</h3><div class="usage"><code>(prot-name uniprot)</code></div><pre class="doc">Returns the name of a uniprot as a string.
</pre></div><div class="public" id="var-subcellular-location"><h3>subcellular-location</h3><div class="usage"><code>(subcellular-location uniprot)</code></div><pre class="doc">Returns a list of maps specifying the sub-cellualr location of a uniprot. Eahc map has the keys - :comments, :orientation, :topology, :location and :evidence. All of which are strings.
</pre></div><div class="public" id="var-wget-uniprot-search"><h3>wget-uniprot-search</h3><div class="usage"><code>(wget-uniprot-search term)</code><code>(wget-uniprot-search term email)</code><code>(wget-uniprot-search term email offset)</code></div><pre class="doc">Returns a 'semi-lazy' (in that 1000 matches are retrieved at time) list of uniprot accession numbers satisfying the search term. The search term uses the same syntax as the uniprot web interface. For example:
   - to get all Schistosoma mansoni proteins in the proteome reference set term would be:
     'organism:6183 AND keyword:1185'
   - get all Schistosoma mansoni proteins in the proteome set that are intrinsic to the
     membrane:
     'taxonomy:6183 AND keyword:1185 AND go:0031224'
   - get all reviewed human entries:
     'reviewed:yes AND organism:9606'
And so on. Returns an empty list if no matches found. Offset refers to the start of the retrieved results and may be of use if a download fails. Uniprot requires an email so an email can be supplied using the email argument.</pre></div><div class="public" id="var-wget-uniprot-sequence"><h3>wget-uniprot-sequence</h3><div class="usage"><code>(wget-uniprot-sequence accessions retype)</code><code>(wget-uniprot-sequence accessions retype email)</code></div><pre class="doc">Takes a list of accessions and returns a 'semi-lazy' (1000 records are fetched at a time) list of sequences corresponding to the accession numbers. The type of sequence entry is specified by 'retype' and can be either :xml for a uniprotProtein object or :fasta for a fastaSequence, no other values are allowed. Uniprot requires an email address so one should be provided in the 'email' argument.
</pre></div></div></body></html>